Implement full Wasm calling convention in IPInt

danlliu opened the pull request:
#### fb3a2765f7892b78721a609c4d653bcf54d0d2a8
<pre>
Implement full Wasm calling convention in IPInt
<a href="https://bugs.webkit.org/show_bug.cgi?id=280230">https://bugs.webkit.org/show_bug.cgi?id=280230</a>
<a href="https://rdar.apple.com/problem/136536169">rdar://problem/136536169</a>

Reviewed by Yusuke Suzuki.

Previously, IPInt did not support a significant range of return signatures, due to
no stack returns being supported. This patch updates IPInt to support returning on
the stack. Additionally, the argument and return value bytecodes are generated
from `wasmCallingConvention()`, ensuring that it is consistent with the rest of
JSC&apos;s Wasm implementations.

* JSTests/wasm/ipint-tests/ipint-test-call-many-f32.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.f32.f32.f32.f32.f32.f32.f32.f32.f32.f32.result.f32.local.8.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-multiret32.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.i32.const.12.i32.const.13.i32.const.14.i32.const.15.i32.const.16.return.func.export.string_appeared_here.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.f32.const.1.f32.const.2.f32.const.3.f32.const.4.f32.const.5.f32.const.6.f32.const.7.f32.const.8.f32.const.9.f32.const.10.f32.const.11.f32.const.12.f32.const.13.f32.const.14.f32.const.15.f32.const.16.return.async test):
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::IPIntCallee::IPIntCallee):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::addReturnData):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::addArguments):
(JSC::Wasm::IPIntGenerator::addEndToUnreachable):

Canonical link: <a href="https://commits.webkit.org/284188@main">https://commits.webkit.org/284188@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a5180dbbce8afc71eba4a957252a34d2983c384

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68668 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48060 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21327 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/72737 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19813 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55856 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19629 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/54751 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13173 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71735 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43917 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59291 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35216 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40583 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16715 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18170 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/61786 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62536 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17062 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74431 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/67916 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12639 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16315 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62225 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12679 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59370 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62252 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15248 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10207 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/3825 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/89695 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43861 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15887 "Found 5 new JSC stress test failures: wasm.yaml/wasm/stress/cc-infinite-int-glitch.js.wasm-collect-continuously, wasm.yaml/wasm/stress/f32-tuple-jsapi.js.wasm-eager, wasm.yaml/wasm/stress/js-wasm-call-many-return-types-on-stack-no-args.js.wasm-bbq, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-bbq, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-slow-memory (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44935 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46129 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44677 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmCallee.h:
```diff
@@ -483,8 +483,6 @@ class IPIntCallee final : public Callee {
     FixedVector<const TypeDefinition*> m_signatures;
 public:
-    // I couldn't figure out how to stop LLIntOffsetsExtractor.cpp from yelling at me.
-    // So just making these public.
```
If it is not longer necessary, let's make them private

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
```diff
@@ -120,28 +120,47 @@ void FunctionIPIntMetadataGenerator::addLEB128V128Constant(v128_t value, size_t
+        auto loc = returnCC.results[i].location;
+
+        if (loc.isGPR()) {
+            ASSERT(loc.jsr().payloadGPR() < NUM_UINT_GPRS);
```
You can use `ASSERT_UNUSED(NUM_UINT_GPRS, loc.jsr().payloadGPR() < NUM_UINT_GPRS);`

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
```diff
@@ -120,28 +120,47 @@ void FunctionIPIntMetadataGenerator::addLEB128V128Constant(v128_t value, size_t
+            ASSERT(loc.jsr().payloadGPR() < NUM_UINT_GPRS);
+            m_uINTBytecode.append(loc.jsr().payloadGPR());
+        } else if (loc.isFPR()) {
+            ASSERT(fprToIndex(loc.fpr()) < NUM_UINT_FPRS);
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
```diff
@@ -120,28 +120,47 @@ void FunctionIPIntMetadataGenerator::addLEB128V128Constant(v128_t value, size_t
+    constexpr static int NUM_UINT_FPRS = 8;
+
+    UNUSED_VARIABLE(NUM_UINT_GPRS);
+    UNUSED_VARIABLE(NUM_UINT_FPRS);
```
We should add some assertion too like,
```
ASSERT_UNUSED(NUM_UINT_GPRS, wasmCallingConvention().jsrArgs.size() <= NUM_UINT_GPRS);
ASSERT_UNUSED(NUM_UINT_FPRS, wasmCallingConvention().fprArgs.size() <= NUM_UINT_FPRS);
```

And let's drop UNUSED_VARIABLE.

Constellation commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -770,39 +770,50 @@ PartialResult WARN_UNUSED_RETURN IPIntGenerator::addTableCopy(unsigned dstTableI
-    int numFPR = 0;
+    m_metadata->m_argumINTBytecode.reserveInitialCapacity(sig->argumentCount() + 1);
+
+    constexpr static int NUM_ARGUMINT_GPRS = 8;
+    constexpr static int NUM_ARGUMINT_FPRS = 8;
+
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -770,39 +770,50 @@ PartialResult WARN_UNUSED_RETURN IPIntGenerator::addTableCopy(unsigned dstTableI
-            }
+        auto loc = callCC.params[i].location;
+        if (loc.isGPR()) {
+            ASSERT(loc.jsr().payloadGPR() < NUM_ARGUMINT_GPRS);
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -770,39 +770,50 @@ PartialResult WARN_UNUSED_RETURN IPIntGenerator::addTableCopy(unsigned dstTableI
+            ASSERT(loc.jsr().payloadGPR() < NUM_ARGUMINT_GPRS);
+            m_metadata->m_argumINTBytecode.append(loc.jsr().payloadGPR());
+        } else if (loc.isFPR()) {
+            ASSERT(fprToIndex(loc.fpr()) < NUM_ARGUMINT_FPRS);
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
```diff
@@ -770,39 +770,50 @@ PartialResult WARN_UNUSED_RETURN IPIntGenerator::addTableCopy(unsigned dstTableI
+    constexpr static int NUM_ARGUMINT_FPRS = 8;
+
+    UNUSED_VARIABLE(NUM_ARGUMINT_GPRS);
+    UNUSED_VARIABLE(NUM_ARGUMINT_FPRS);
```
Then we do not need this.

Constellation commented on Source/JavaScriptCore/wasm/WasmCallee.h:
```diff
@@ -443,6 +443,10 @@ class IPIntCallee final : public Callee {
 
+    unsigned getNumLocals() const { return m_numLocals; }
+    unsigned getLocalSizeToAlloc() const { return m_localSizeToAlloc; }
+    unsigned getRethrowSlots() const { return m_numRethrowSlotsToAlloc; }
```
Let's drop `get` to align it to WebKit's style.

danlliu committed:
Implement full Wasm calling convention in IPInt
https://bugs.webkit.org/show_bug.cgi?id=280230
rdar://problem/136536169

Reviewed by Yusuke Suzuki.

Previously, IPInt did not support a significant range of return signatures, due to
no stack returns being supported. This patch updates IPInt to support returning on
the stack. Additionally, the argument and return value bytecodes are generated
from `wasmCallingConvention()`, ensuring that it is consistent with the rest of
JSC's Wasm implementations.

* JSTests/wasm/ipint-tests/ipint-test-call-many-f32.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.param.f32.f32.f32.f32.f32.f32.f32.f32.f32.f32.result.f32.local.8.return.async test):
* JSTests/wasm/ipint-tests/ipint-test-multiret32.js: Added.
(from.string_appeared_here.import.as.assert.from.string_appeared_here.let.wat.module.func.export.string_appeared_here.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.result.i32.i32.const.1.i32.const.2.i32.const.3.i32.const.4.i32.const.5.i32.const.6.i32.const.7.i32.const.8.i32.const.9.i32.const.10.i32.const.11.i32.const.12.i32.const.13.i32.const.14.i32.const.15.i32.const.16.return.func.export.string_appeared_here.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.result.f32.f32.const.1.f32.const.2.f32.const.3.f32.const.4.f32.const.5.f32.const.6.f32.const.7.f32.const.8.f32.const.9.f32.const.10.f32.const.11.f32.const.12.f32.const.13.f32.const.14.f32.const.15.f32.const.16.return.async test):
* Source/JavaScriptCore/llint/InPlaceInterpreter.asm:
* Source/JavaScriptCore/llint/InPlaceInterpreter.h:
* Source/JavaScriptCore/llint/InPlaceInterpreter64.asm:
* Source/JavaScriptCore/wasm/WasmCallee.cpp:
(JSC::Wasm::IPIntCallee::IPIntCallee):
* Source/JavaScriptCore/wasm/WasmCallee.h:
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.cpp:
(JSC::Wasm::FunctionIPIntMetadataGenerator::addReturnData):
* Source/JavaScriptCore/wasm/WasmFunctionIPIntMetadataGenerator.h:
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntGenerator::addArguments):
(JSC::Wasm::IPIntGenerator::addEndToUnreachable):

Canonical link: https://commits.webkit.org/284188@main
